src/manifest.c

```diff
@@ -298,25 +298,35 @@ static void set_untracked_manifest_files(struct manifest *manifest)
 }
 
 /* Removes the extracted Manifest.<bundle> and accompanying tar file, cache file, and
- * the signature file */
+ * the signature file from the statedir and statedir-cache */
 static void remove_manifest_files(char *filename, int version, char *hash)
 {
 	char *file;
+	char *state_dirs[] = { globals.state_dir, globals.state_dir_cache };
```

Comment:
When the statedir contains a corrupted manifest, the statedir-cache will also be wiped, so it cannot be used as a fallback. It would be better for the statedir-cache to be used as a fallback in this case, but the current fallback code for manifests doesn't work well for more than 1 fallback method.

The manifest fallback code could be refactored to better support the statedir-cache. The `retrieve_manifests` function could validate manifest hashes instead of only checking for the existence of correctly named manifests (this is how the fullfile fallback code works). This would make it easy to find the desired manifest by checking the statedir, checking the statedir-cache, or downloading over the network.